pkg/controllers/observability-ui/observability-ui-plugin/controller.go

```diff
@@ -0,0 +1,218 @@
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	obsui "github.com/rhobs/observability-operator/pkg/apis/observability-ui/v1alpha1"
+	"github.com/rhobs/observability-operator/pkg/reconciler"
+)
+
+type resourceManager struct {
+	k8sClient      client.Client
+	scheme         *runtime.Scheme
+	logger         logr.Logger
+	controller     controller.Controller
+	pluginConf     UIPluginsConfiguration
+	clusterVersion string
+}
+
+type UIPluginsConfiguration struct {
+	Images map[string]string
+}
+
+type Options struct {
+	PluginsConf UIPluginsConfiguration
+}
+
+// RBAC for managing UIPlugins
+// +kubebuilder:rbac:groups=observability.openshift.io,resources=uiplugins,verbs=get;list;watch;create;update;delete;patch
+// +kubebuilder:rbac:groups=observability.openshift.io,resources=uiplugins/status,verbs=get;update
+
+// RBAC for managing observability ui plugin objects
+//+kubebuilder:rbac:groups=apps,resources=deployments,verbs=list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=serviceaccounts;services;configmaps,verbs=get;list;watch;create;update;patch;delete
+
+// RBAC for managing Console CRs
+// +kubebuilder:rbac:groups=operator.openshift.io,resources=consoles,verbs=get;patch;list;watch
+// +kubebuilder:rbac:groups=console.openshift.io,resources=consoleplugins,verbs=get;list;watch;create;update;delete;patch
+
+// RBAC for reading cluster version
+// +kubebuilder:rbac:groups=config.openshift.io,resources=clusterversions,verbs=get;list;watch
+
+func RegisterWithManager(mgr ctrl.Manager, opts Options) error {
+	logger := ctrl.Log.WithName("observability-ui")
+
+	clusterVersion, err := getClusterVersion(mgr.GetAPIReader())
+
+	if err != nil {
+		logger.Error(err, "failed to get cluster version")
+		return err
+	}
+
+	rm := &resourceManager{
+		k8sClient:      mgr.GetClient(),
+		scheme:         mgr.GetScheme(),
+		logger:         logger,
+		pluginConf:     opts.PluginsConf,
+		clusterVersion: clusterVersion.Status.Desired.Version,
+	}
+
+	generationChanged := builder.WithPredicates(predicate.GenerationChangedPredicate{})
+
+	ctrl, err := ctrl.NewControllerManagedBy(mgr).
+		For(&obsui.UIPlugin{}).
+		Owns(&appsv1.Deployment{}, generationChanged).
+		Owns(&v1.Service{}, generationChanged).
+		Owns(&v1.ServiceAccount{}, generationChanged).
+		Owns(&rbacv1.ClusterRole{}, generationChanged).
+		Owns(&rbacv1.ClusterRoleBinding{}, generationChanged).
+		Owns(&osv1alpha1.ConsolePlugin{}, generationChanged).
+		Build(rm)
+	if err != nil {
+		return err
+	}
+	rm.controller = ctrl
+
+	return nil
+}
+
+func getClusterVersion(k8client client.Reader) (*configv1.ClusterVersion, error) {
+	clusterVersion := &configv1.ClusterVersion{}
+	key := client.ObjectKey{Name: "version"}
+	if err := k8client.Get(context.TODO(), key, clusterVersion); err != nil {
+		return nil, err
+	}
```

Comment:
Double checking here. Is a NotFound error also worth to return and in turn reconcile once again? Usually NotFound errors are treated as non-queuing. WDYT?

Reply:
This function gets called before the controller is created so IIUC is not part of the reconciliation. Meaning, if getting the cluster version fails, the controller does not register with the manager.